File: BeerViewer/Models/Raw/Notifier.cs
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace BeerViewer.Models.Raw
{
	public partial class Notifier : INotifyPropertyChanged
	{
		public event PropertyChangedEventHandler PropertyChanged;

		protected virtual void RaisePropertyChanged([CallerMemberName] string propertyName = null)
		{
			this.PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
		}

		public void PropertyEvent(string PropertyName, Action Handler, bool RaiseRegistered = false)
		{
			this.PropertyChanged += (_, e) =>
			{
				if (PropertyName == e.PropertyName)
					Handler?.Invoke();
			};

			if (RaiseRegistered)
				Handler?.Invoke();
		}
	}
}

File: WindowFrame/source/System/Base/Notifier.ts
interface PropertyChanged {
    (propertyName: string | null): void;
}
interface INotifyPropertyChanged {
    PropertyChanged: PropertyChanged | PropertyChanged[] | null;
}

class Notifier implements INotifyPropertyChanged {
    public PropertyChanged: PropertyChanged | PropertyChanged[] | null = null;

    protected RaisePropertyChanged(propertyName: string | null = null): void {
        if (this.PropertyChanged === null) return;

        if ((<PropertyChanged[]>this.PropertyChanged).length) {
            const list = this.PropertyChanged as PropertyChanged[];
            list.forEach(x => x(propertyName));
            return;
        }
        (<PropertyChanged>this.PropertyChanged)(propertyName);
    }

    public PropertyEvent(PropertyName: string, Handler: Function, RaiseRegistered: boolean = false): void {
        if (this.PropertyChanged === null) {
            this.PropertyChanged = x => {
                if (x === PropertyName && Handler)
                    Handler();
            };
        }

        if (RaiseRegistered && Handler) Handler();
    }
}
class TickNotifier extends Notifier implements IDisposable {
    private handler: number;

    constructor() {
        super();
        this.handler = setInterval(this.Tick, 1000);
    }

    protected Tick(): void { }

    Dispose(): void {
        clearInterval(this.handler);
    }
}